Build error with GCC 8.x.y

**Describe the bug**
Hi all.
I have tried to build project on my local machine but it doesn't compile.

**To Reproduce**
```
$ cmake -Wdev -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DBOOST_LIBRARYDIR=G:\\_LIBS\\boost_1_70_0\\stage\\lib -DBoost_INCLUDE_DIR=G:\\_LIBS\\boost_1_70_0 -DBOOST_ROOT=G:\\_LIBS\\boost_1_70_0\\boost -DCMAKE_PREFIX_PATH=C:\\Qt\\Qt5.12.0\\5.15.0\\mingw81_64\\bin -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ ..
-- The C compiler identification is GNU 8.1.0
-- The CXX compiler identification is GNU 8.1.0
-- Check for working C compiler: C:/Qt/Qt5.12.0/Tools/mingw810_64/bin/gcc.exe
-- Check for working C compiler: C:/Qt/Qt5.12.0/Tools/mingw810_64/bin/gcc.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Qt/Qt5.12.0/Tools/mingw810_64/bin/g++.exe
-- Check for working CXX compiler: C:/Qt/Qt5.12.0/Tools/mingw810_64/bin/g++.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using compiler: C:/Qt/Qt5.12.0/Tools/mingw810_64/bin/g++.exe
-- Boost static libs: ON
-- Boost multithreaded libs: ON
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Using compiler: C:/Qt/Qt5.12.0/Tools/mingw810_64/bin/g++.exe
-- Boost static libs: ON
-- Boost multithreaded libs: ON
-- Found Boost: G:\_LIBS\boost_1_70_0 (found version "1.70.0") found components: filesystem system thread chrono date_time atomic
-- x64 configuration
-- Degate version: 2.1.0-alpha.1
-- Release date: unreleased
-- Doxygen not found
-- Configuring done
-- Generating done
-- Build files have been written to: G:/personal_projects/Degate/rel_win_build

$ make -j VERBOSE=1
```

**Expected behavior**
Project should be compiled without any error.

**Screenshots**
```
In file included from C:/Qt/Qt5.12.0/Tools/mingw810_64/lib/gcc/x86_64-w64-mingw32/8.1.0/include/c++/map:60,
                 from G:/personal_projects/Degate/src/Globals.h:29,
                 from G:/personal_projects/Degate/src/Core/Image/Image.h:45,
                 from G:/personal_projects/Degate/src/Core/LogicModel/PlacedLogicModelObject.h:26,
                 from G:/personal_projects/Degate/src/Core/LogicModel/ObjectSet.h:25,
                 from G:/personal_projects/Degate/src/Core/LogicModel/ObjectSet.cc:24:
C:/Qt/Qt5.12.0/Tools/mingw810_64/lib/gcc/x86_64-w64-mingw32/8.1.0/include/c++/bits/stl_tree.h: In instantiation of 'class std::_Rb_tree<std::shared_ptr<degate::GatePort>, std::shared_ptr<degate::GatePort>, std::_Identity<std::shared_ptr<degate::GatePort> >, degate::LMOCompare, std::allocator<std::shared_ptr<degate::GatePort> > >':
C:/Qt/Qt5.12.0/Tools/mingw810_64/lib/gcc/x86_64-w64-mingw32/8.1.0/include/c++/bits/stl_set.h:133:17:   required from 'class std::set<std::shared_ptr<degate::GatePort>, degate::LMOCompare>'
G:/personal_projects/Degate/src/Core/LogicModel/Gate/Gate.h:60:53:   required from here
C:/Qt/Qt5.12.0/Tools/mingw810_64/lib/gcc/x86_64-w64-mingw32/8.1.0/include/c++/bits/stl_tree.h:452:21: error: static assertion failed: comparison object must be invocable with two arguments of key type
       static_assert(__is_invocable<_Compare&, const _Key&, const _Key&>{},
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make.exe[2]: *** [CMakeFiles/DegateCore.dir/src/Core/LogicModel/ObjectSet.cc.obj] Error 1
make.exe[2]: Leaving directory `G:/personal_projects/Degate/win_build'
make.exe[1]: *** [CMakeFiles/DegateCore.dir/all] Error 2
make.exe[1]: Leaving directory `G:/personal_projects/Degate/win_build'
make.exe: *** [all] Error 2
```

**Desktop (please complete the following information):**
 - OS: Win7 x64
 - Version: 2.1.0-alpha.1
 - GCC (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0
 - Qt 5.15.0

**Additional context**
[Here](https://bugzilla.redhat.com/show_bug.cgi?id=1541688) seems to be the same error.

--- a/src/Core/LogicModel/LogicModelObjectBase.cc
+++ b/src/Core/LogicModel/LogicModelObjectBase.cc
@@ -121,14 +121,3 @@ const std::string LogicModelObjectBase::get_object_type_name() const
 {
     return tr("Generic object").toStdString();
 }
-
-
-bool LMOCompare::operator()(const LogicModelObjectBase& a, const LogicModelObjectBase& b) const
-{
-    return a.get_object_id() < b.get_object_id();
-}
-
-bool LMOCompare::operator()(const LogicModelObjectBase_shptr& a, const LogicModelObjectBase_shptr& b) const
-{
-    return this->operator()(*a, *b);
-}

--- a/src/Core/LogicModel/LogicModelObjectBase.h
+++ b/src/Core/LogicModel/LogicModelObjectBase.h
@@ -136,19 +136,28 @@ namespace degate
         virtual const std::string get_object_type_name() const;
     };
 
-    typedef std::shared_ptr<LogicModelObjectBase> LogicModelObjectBase_shptr;
-
+    /**
+     * @class LMOCompare
+     * @brief Helper to compare LogicModelObjectBase.
+     *
+     * Used with std::set.
+     */
     class LMOCompare
     {
     public:
         template <typename T>
         bool operator()(const T& a, const T& b) const
         {
+            static_assert(std::is_base_of<LogicModelObjectBase, T>::value, "T must inherit from LogicModelObjectBase.");
+
             return a.get_object_id() < b.get_object_id();
         }
+
         template <typename T>
         bool operator()(const std::shared_ptr<T>& a, const std::shared_ptr<T>& b) const
         {
+            static_assert(std::is_base_of<LogicModelObjectBase, T>::value, "T must inherit from LogicModelObjectBase.");
+
             return this->operator()(*a, *b);
         }
     };

--- a/src/Core/LogicModel/LogicModelObjectBase.h
+++ b/src/Core/LogicModel/LogicModelObjectBase.h
@@ -141,8 +141,16 @@ namespace degate
     class LMOCompare
     {
     public:
-        bool operator()(const LogicModelObjectBase& a, const LogicModelObjectBase& b) const;
-        bool operator()(const LogicModelObjectBase_shptr& a, const LogicModelObjectBase_shptr& b) const;
+        template <typename T>
+        bool operator()(const T& a, const T& b) const
+        {
+            return a.get_object_id() < b.get_object_id();
+        }
+        template <typename T>
+        bool operator()(const std::shared_ptr<T>& a, const std::shared_ptr<T>& b) const
+        {
+            return this->operator()(*a, *b);
+        }
     };
 }
 

